### PR TITLE
Bump MSRV to Rust version 1.56.1

### PR DESCRIPTION
**Pull request by tcharding**
Rust version 1.56.0 introduced edition 2021. Shortly afterwards, on October 21 2021 Rust version 1.56.1 was released.

Debian stable is currently shipping `rustc 1.63.0`. Our stated MSRV policy is: In Debian stable and at least 2 years old. Therefore our MSRV policy is met by Rust version 1.56.1 and we can strat to bump our MSRV org wide. Start by bumping the `rust-bitcoin` and `hashes` MSRV to Rust 1.56.1

Start by bumping the `rust-bitcoin` and `hashes` MSRV to Rust 1.56.1, includes:
    
- Update docs.
- Update CI and remove pinning.
- Update the build files and remove now stale cfg attributes rust_v_1_x for values less than the new MSRV.
- Use new `IntoIterator` for arrays so we no longer need to allocate a vector to iterate.

Links:

- https://blog.rust-lang.org/2021/11/01/Rust-1.56.1.html
- https://blog.rust-lang.org/2021/10/21/Rust-1.56.0.html
- https://packages.debian.org/stable/rust/rustc

### CODE REVIEW

File: bitcoin/src/taproot.rs [hunk 1]
```diff
@@ -1842,7 +1842,7 @@ mod test {
         let tree_info = builder.finalize(&secp, internal_key).unwrap();
         let output_key = tree_info.output_key();
 
-        for script in vec![a, b, c, d, e] {
+        for script in [a, b, c, d, e] {
```
Comment (Kixunil) [inline bitcoin/src/taproot.rs:1845]: Oh, yeah, `IntoIterator` in arrays! :tada: :tada: :tada:

**Review by Kixunil — COMMENTED** [review]
While technically correct, it keeps `MSRV_MINOR` at 48 in `build.rs`. It might be nicer to bump it and remove `#[cfg(rust_*)]` for versions less than or equal to 1.56.

Further, this version stabilized the `rust-version` field in `Cargo.toml` and we should set it. It might reduce the burden of bumping the numbers in the future if we can pull them from that field. The release notes of Rust 1.56 specifically say "we encourage CI systems
  that support Rust code to include a crate's specified minimum version in the test matrix for that crate by default." but I didn't check if our CI supports it.

I think doing these makes sense in this PR but wouldn't block it.

**Comment by tcharding** [comment]
Implemented suggestions and squashed it all into a single patch, re-wrote git log and PR description to match new patch.

**Review by Kixunil — DISMISSED** [review]
ACK e729589955bb2b39c5178f255fdd40c3b68b46ee

**Comment by tcharding** [comment]
Oh, sorry man was working while you reviewed. I added a patch to update to edition 2021.

**Review by Kixunil — DISMISSED** [review]
ACK 4ab2c2889630a150bd02328623af2218a36d5aa5

**Comment by Kixunil** [comment]
Oh, I just notice you didn't set `rust-version` in `Cargo.toml`. Was it on purpose?

**Comment by tcharding** [comment]
Not on purpose, just an omission on my part. Will add

**Review by Kixunil — DISMISSED** [review]
ACK 5afcc2b8d5cb6453eb680cc8b6f2cacc4667ede4

**Comment by apoelstra** [comment]
cc @TheBlueMatt @stevenroose I'll give you two a week to comment or veto this, or else I'll merge it on Wednesday Nov 22.

**Review by apoelstra — APPROVED** [review]
ACK 5afcc2b8d5cb6453eb680cc8b6f2cacc4667ede4

**Comment by apoelstra** [comment]
@TheBlueMatt yep, for sure. We're in no particular rush.

**Comment by Kixunil** [comment]
Well it won't be released right now anyway so it shouldn't matter? Or is that user using our `master` branch?

**Comment by tcharding** [comment]
Thanks man! Rebased.

**Comment by apoelstra** [comment]
Might as well do it now then, since we don't have a major release planned til April.

**Review by apoelstra — APPROVED** [review]
ACK dda8765827e6d8807f4cac8e5186a20348718d64

**Comment by tcharding** [comment]
Rebased, no other changes.

**Review by apoelstra — DISMISSED** [review]
ACK dda8765827e6d8807f4cac8e5186a20348718d64

File: Cargo.toml [hunk 2]
```diff
@@ -1,5 +1,6 @@
 [workspace]
```
Comment (Kixunil) [inline Cargo.toml:1]: We should be able to remove all `use` statements importing `TryFrom` and `TryInto` now. Fine to do in as separate PR.

**Review by Kixunil — APPROVED** [review]
ACK 4d5415f83545bba238f5b78b1730bfb14e7d86ba

File: Cargo.toml (hunk 2 above)
Comment (tcharding) [reply Cargo.toml:1]: Unusualy, clippy told me about a few instances of `TryInto` but all of them. And none of the `TryFrom`. I opened some files to play with clippy and ended up just doing the patch to remove them all. Its on top.

**Review by Kixunil — APPROVED** [review]
ACK 761de886be0be2b955126381a87581166b642ccc

**Review by apoelstra — APPROVED** [review]
ACK 761de886be0be2b955126381a87581166b642ccc

File: contrib/test.sh [hunk 3]
```diff
@@ -4,22 +4,6 @@ set -ex
 
 CRATES="bitcoin hashes internals fuzz"
 DEPS="recent minimal"
-MSRV="1\.48\.0"
-
-# Test pinned versions.
-if cargo --version | grep ${MSRV}; then
-    cargo update -p serde_json --precise 1.0.99
```
Comment (yancyribbens) [inline contrib/test.sh:11]: Nice!

File: contrib/test.sh [hunk 4]
```diff
@@ -4,22 +4,6 @@ set -ex
 
 CRATES="bitcoin hashes internals fuzz"
 DEPS="recent minimal"
-MSRV="1\.48\.0"
-
-# Test pinned versions.
-if cargo --version | grep ${MSRV}; then
```
Comment (yancyribbens) [inline contrib/test.sh:10]: Nice!

**Review by apoelstra — APPROVED** [review]
ACK 761de886be0be2b955126381a87581166b642ccc
